ui/js/actions/settings.js

```diff
@@ -41,3 +44,296 @@ export function doSetClientSetting(key, value) {
+    tl: ["Tagalog / Filipino", "Tagalog"],
+    tn: ["Tswana", "Setswana"],
+    to: ["Tonga", "Lea Faka-Tonga"],
+    tr: ["Turkish", "Türkçe"],
+    ts: ["Tsonga", "Xitsonga"],
+    tt: ["Tatar", "Tatarça"],
+    tw: ["Twi", "Twi"],
+    ty: ["Tahitian", "Reo Mā`ohi"],
+    ug: ["Uyghur", "Uyƣurqə / ئۇيغۇرچە"],
+    uk: ["Ukrainian", "Українська"],
+    ur: ["Urdu", "اردو"],
+    uz: ["Uzbek", "Ўзбек"],
+    ve: ["Venda", "Tshivenḓa"],
+    vi: ["Vietnamese", "Tiếng Việt"],
+    vo: ["Volapük", "Volapük"],
+    wa: ["Walloon", "Walon"],
+    wo: ["Wolof", "Wollof"],
+    xh: ["Xhosa", "isiXhosa"],
+    yi: ["Yiddish", "ייִדיש"],
+    yo: ["Yoruba", "Yorùbá"],
+    za: ["Zhuang", "Cuengh / Tôô / 壮语"],
+    zh: ["Chinese", "中文"],
+    zu: ["Zulu", "isiZulu"],
+  };
+
+  const lang = locale.substring(0, 2);
+  return {
+    type: types.LANGUAGE_RESOLVED,
+    data: { key: locale, value: `${langs[lang][0]} (${langs[lang][1]})` },
+  };
+}
+
+export function doDownloadLanguage(lang, destinationPath) {
+  return function(dispatch, getState) {
+    const plainLanguage = lang.replace(".json", "");
+
+    dispatch({
+      type: types.DOWNLOAD_LANGUAGE_STARTED,
+      data: plainLanguage,
+    });
+
+    return new Promise((resolve, reject) => {
+      const file = fs.createWriteStream(destinationPath);
+      const req = http.request(
+        `http://i18n.lbry.io/langs/${lang}`,
+        response => {
+          file.on("finish", () => {
+            file.close();
+
+            // push to our local list
+            dispatch({
+              type: types.DOWNLOAD_LANGUAGE_SUCCEEDED,
+              data: plainLanguage,
+            });
+
+            resolve();
+          });
+
+          response.pipe(file);
+        }
+      );
+
+      const errorHandler = err => {
+        if (file) {
+          file.close();
+        }
+
+        fs.unlink(destinationPath); // Delete the file async. (But we don't check the result)
+        dispatch({
+          type: types.DOWNLOAD_LANGUAGE_FAILED,
+          data: plainLanguage,
+        });
+        reject(err);
+      };
+      req.on("error", errorHandler);
+      file.on("error", errorHandler);
+
+      req.end();
+    });
+  };
+}
+
+export function doDownloadLanguages() {
+  return function(dispatch, getState) {
+    if (!fs.existsSync("app/locales")) {
+      fs.mkdirSync("app/locales");
+    }
+
+    dispatch({
+      type: types.DOWNLOAD_LANGUAGES_STARTED,
+      data: {},
+    });
+
+    return new Promise((resolve, reject) => {
+      const req = http.request(
+        { host: "i18n.lbry.io", path: "/" },
+        response => {
+          let str = "";
+
+          response.on("data", chunk => {
+            str += chunk;
+          });
+
+          response.on("end", () => {
+            const files = JSON.parse(str);
+            const actions = [];
+            files.forEach(file => {
+              actions.push(doDownloadLanguage(file, `app/locales/${file}`));
```

Comment:
Based on the original logic, it looks like the downloaded language files are used to determine the languages that are displayed on the Settings page (except English which is the default). I feel changing the display logic here to make use the language index downloaded from http://i18n.lbry.io/ could be done in another pull request.

Also, if we have to download the language on demand, I suppose we have to have some sort of wait dialog (while the file downloads) on the Settings page before the language change is committed.